src/core/ddsi/src/ddsi_typelookup.c

```diff
@@ -0,0 +1,356 @@
+      pp_found = true;
+  }
+  entidx_enum_participant_fini (&est);
+  return pp;
+}
+
+bool ddsi_tl_request_type (struct ddsi_domaingv * const gv, const type_identifier_t *type_id)
+{
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
+  struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, type_id);
+  if (tlm->state != TL_META_NEW)
+  {
+    GVTRACE ("ddsi_tl_request_type - state not-new (%u) for "PTYPEIDFMT"\n", tlm->state, PTYPEID (*type_id));
+    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+    return false;
+  }
+
+  struct participant *pp = get_typelookup_pp (gv, &tlm->dst_prefix);
+  struct writer *wr = get_builtin_writer (pp, NN_ENTITYID_TL_SVC_BUILTIN_REQUEST_WRITER);
+  if (wr == NULL)
+  {
+    GVTRACE ("ddsi_tl_request_type("PGUIDFMT") - builtin TypeLookup service request writer not found\n", PGUID (pp->e.guid));
+    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+    return false;
+  }
+
+  type_lookup_request_t request;
+  memcpy (&request.writer_guid, &wr->e.guid, sizeof (wr->e.guid));
+  request.sequence_number = wr->seq;
+  request.type_ids.n = 1;
+  request.type_ids.type_ids = tlm->type_id;
+  struct ddsi_serdata *serdata = ddsi_serdata_from_sample (gv->tl_svc_request_type, SDK_DATA, &request);
+  serdata->timestamp = ddsrt_time_wallclock ();
+
+  GVTRACE (" wr-tl-req "PGUIDFMT" typeid "PTYPEIDFMT"\n", PGUID (wr->e.guid), PTYPEID(*type_id));
+  struct ddsi_tkmap_instance *tk = ddsi_tkmap_lookup_instance_ref (gv->m_tkmap, serdata);
+  write_sample_gc (lookup_thread_state (), NULL, wr, serdata, tk);
+  tlm->state = TL_META_REQUESTED;
+  ddsi_tkmap_instance_unref (gv->m_tkmap, tk);
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+  return true;
+}
+
+static void write_typelookup_reply (struct participant *pp, seqno_t seqno, type_identifier_type_object_pair_seq_t *types)
+{
+  struct ddsi_domaingv * const gv = pp->e.gv;
+  type_lookup_reply_t reply;
+
+  struct writer *wr = get_builtin_writer (pp, NN_ENTITYID_TL_SVC_BUILTIN_REPLY_WRITER);
+  if (wr == NULL)
+  {
+    GVTRACE ("write_typelookup_reply("PGUIDFMT") - builtin TypeLookup service reply writer not found\n", PGUID (pp->e.guid));
+    return;
+  }
+
+  memcpy (&reply.writer_guid, &wr->e.guid, sizeof (wr->e.guid));
+  reply.sequence_number = seqno;
+  reply.types.n = types->n;
+  reply.types.types = types->types;
+  struct ddsi_serdata *serdata = ddsi_serdata_from_sample (gv->tl_svc_reply_type, SDK_DATA, &reply);
+  serdata->timestamp = ddsrt_time_wallclock ();
+
+  GVTRACE (" wr-tl-reply wr "PGUIDFMT"\n", PGUID (wr->e.guid));
+  struct ddsi_tkmap_instance *tk = ddsi_tkmap_lookup_instance_ref (gv->m_tkmap, serdata);
+  write_sample_gc (lookup_thread_state (), NULL, wr, serdata, tk);
+  ddsi_tkmap_instance_unref (gv->m_tkmap, tk);
+}
+
+void ddsi_tl_handle_request (struct ddsi_domaingv *gv, const ddsi_guid_prefix_t *guid_prefix, struct ddsi_serdata *sample_common)
+{
+  assert (!(sample_common->statusinfo & (NN_STATUSINFO_DISPOSE | NN_STATUSINFO_UNREGISTER)));
```

Comment:
I haven't checked yet, but in case I forget: if this is an assert, the caller should ensure that this precondition is met before calling this function.

Reply:
I moved this to `q_ddsi_discovery` as a condition for calling `ddsi_tl_handle_request` and `ddsi_tl_handle_request`